src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceDAO.scala

```diff
@@ -37,4 +43,25 @@ class FileSystemWorkspaceDAO(storageDirectory: Path) extends WorkspaceDAO {
     val json = new String(Files.readAllBytes(location)).parseJson
     WorkspaceFormat.read(json)
   }
+
+  def list(): Seq[WorkspaceShort] = {
+    val visitor = new WorkspaceFileVisitor(storageDirectory)
+    Files.walkFileTree(storageDirectory, visitor)
+    visitor.workspaces.result
+  }
 }
+
+private class WorkspaceFileVisitor(root: Path) extends SimpleFileVisitor[Path] with LazyLogging {
```

Comment:
I feel like whenever you find yourself reaching into GoF there's probably a better way to do things (a la the notion that design patterns make up for missing language features). Like here, it's really just a foreach() across the collection of all files in the dir tree.

So this bothers me but in this particular case I think it's fine since we already have SimpleFileVisitor which already codes up the recursive dir tree generation. But it'd be nice if we could say something like:

     getAllFilesRecursively(path) foreach { doSomething() }

tl;dr - this disturbs my inner FP-fan enough to speak up but you should probably leave it as is